Source/Painting/SvgPatternServer.cs

```diff
@@ -163,71 +157,84 @@ public override Brush GetBrush(SvgVisualElement renderingElement, ISvgRenderer r
+                        x *= bounds.Width;
+                    if (yUnit.Type != SvgUnitType.Percentage)
+                        y *= bounds.Height;
+                    if (widthUnit.Type != SvgUnitType.Percentage)
+                        width *= bounds.Width;
+                    if (heightUnit.Type != SvgUnitType.Percentage)
+                        height *= bounds.Height;
+                    x += bounds.X;
+                    y += bounds.Y;
+                }
 
-                    var image = new Bitmap((int)Math.Ceiling(width), (int)Math.Ceiling(height));
-                    using (var iRenderer = SvgRenderer.FromImage(image))
+                var tile = new Bitmap((int)Math.Ceiling(width), (int)Math.Ceiling(height));
+                using (var tileRenderer = SvgRenderer.FromImage(tile))
+                {
+                    tileRenderer.SetBoundable(renderingElement);
+                    if (viewBox != SvgViewBox.Empty)
+                    {
+                        var bounds = tileRenderer.GetBoundable().Bounds;
+                        tileRenderer.ScaleTransform(width / viewBox.Width, height / viewBox.Height);
+                    }
+                    else if (patternContentUnits == SvgCoordinateUnits.ObjectBoundingBox)
                     {
-                        iRenderer.SetBoundable((_patternContentUnits == SvgCoordinateUnits.ObjectBoundingBox) ? new GenericBoundable(0, 0, width, height) : renderer.GetBoundable());
-                        iRenderer.Transform = patternMatrix;
-                        iRenderer.SmoothingMode = SmoothingMode.AntiAlias;
-                        iRenderer.SetClip(new Region(new RectangleF(0, 0,
-                            viewBox.Width > 0 ? viewBox.Width : width,
-                            viewBox.Height > 0 ? viewBox.Height : height)));
-
-                        foreach (var child in childElem.Children)
-                            child.RenderElement(iRenderer);
+                        var bounds = tileRenderer.GetBoundable().Bounds;
+                        tileRenderer.ScaleTransform(bounds.Width, bounds.Height);
                     }
 
-                    var textureBrush = new TextureBrush(image, new RectangleF(0f, 0f, width, height));
-                    var brushTransform = EffectivePatternTransform.Clone();
-                    brushTransform.Translate(x, y, MatrixOrder.Append);
-                    textureBrush.Transform = brushTransform;
-                    return textureBrush;
+                    foreach (var child in hasChildren.Children)
+                        child.RenderElement(tileRenderer);
                 }
+
+                var textureBrush = new TextureBrush(tile, new RectangleF(0f, 0f, width, height))
+                {
+                    Transform = EffectivePatternTransform
+                };
+                textureBrush.TranslateTransform(x, y);
+                return textureBrush;

```

Comment:
This all looks very clean now, though I have trouble understanding where exactly the problem was in the old code. Can you elaborate a bit? (just out of interest)

Reply:
@mrbean-bremen 

It looks like a big change because there are many differences, but process flow has not changed.

Main change is:

- Unify `unit` (`patternUnit` and `patternContentUnit`).